Proposed changes to Mpd2 formatter method

MagusLucius opened the pull request:
The proposed changes fix the following bugs:
    1) logger spamming error messages once per update when tags are missing.
    2) widget crashing (forcing qtile restart to fix) when playlist is empty.

The changes add the following functionality:
    1) a new 'remaining' tag, which is 'fuilltime' - 'elapsed'
    2) a new system for introducing user formatting methods for each tag.

flacjacket commented on libqtile/widget/mpd2widget.py:
```diff
@@ -165,28 +187,59 @@ def __getitem__(self, key):
+        del song_info['time']
+
+        song_info.update(status)
+        if song_info['updating_db'] == default:
```
You can say here `if 'updating_db' not in song_info`

flacjacket commented on libqtile/widget/mpd2widget.py:
```diff
@@ -165,28 +187,59 @@ def __getitem__(self, key):
             return "ERROR"
 
-    def prepare_formatting(self, status, currentsong):
+    # Removed an unused parameter from this function.
```
You can remove this comment

flacjacket commented on libqtile/widget/mpd2widget.py:
```diff
@@ -165,28 +187,59 @@ def __getitem__(self, key):
+        # Remaining should default to '00:00' if either or both are missing.
+        if 'remaining' in self.status_format:
+            total = float(song_info['fulltime'])\
+                if song_info['fulltime'] != default else 0.0
```
These as well should use eg `'fulltime' in song_info`

flacjacket commented on libqtile/widget/mpd2widget.py:
```diff
@@ -165,28 +187,59 @@ def __getitem__(self, key):
-        status.update(currentsong)
+    def formatter(self, status, current_song):
+        default = 'Undefined'
+        song_info = defaultdict(lambda: default)
```
It doesn't look like anything here is necessarily needing the defaultdict, rather than a normal dict. What properties do you need to have added to keep the formatting from failing? Maybe you can conditionally add them at the end?

MagusLucius replied:
I apparently have quite a number of corrupted tags in my library, all mp3 files from what I can tell.  MPD refuses to read the tags because of the presence of an invalid frame.  When this happens, the currentsong() method returns an empty dict.  Originally, I was using regex to fish the requested tags out of the format string, but I thought using a defaultdict was a better idea, as it would return a default value for any missing tag, without the overhead of a regular expression.

MagusLucius replied:
I've thought of a few more edits, but I haven't pushed them to my fork yet.
1) For the sake of consistency, I moved the parameters 'prepare_status' and 'status_format' out of the constructor, and put them in the defaults dictionary instead.  I see that's the way the old mpd widget handled things.

2) I added an Idle state message that replaces the regular message when MPD is idle (e.g. status['state'] == 'stop' and current_song == {}) There's a secondary format string that I set to default to "{play_status} {idle_message} [{repeat}{random}{single}{consume}{updating_db}]"

3) I added a default formatting function ('all': lambda s: cgi.escape(s)) which prevents pango from throwing a fit if you use certain characters like "&" (I use this character because I'm using a special font to display the status symbols and the "shuffle" glyph is mapped to "&".  The character also shows up occasionally in artist information for songs. The 'all' function gets applied before any other user formatting.  I'm not sure if this is a great idea or not (partially why I haven't updated the fork) the presence of 'all' in the formatting function dict means that a user would then have to take that function's result into account with every other formatting function in the dictionary. This isn't difficult in the case of a simple escape sequence (I use ast.literal_eval to convert the escaped numbers back into numbers, but it's an extra step).  I did note that in my documentation, however.

let me know what you think.

ramnes commented:
@MagusLucius can you rebase on master please?

MagusLucius commented:
> @MagusLucius can you rebase on master please?

Done.  The reason I switched it to develop was that the README said to make PR's against develop.

MagusLucius commented:
Changed mouse interaction method for more flexibility.
Added configurable Idle message for when playlist is empty
Fixed pep8 errors (including 2247.7 and 2247.8)

MagusLucius commented:
I edited to add mpd info to the docs and sandboxed the widget internals access for mouse_button, but it isn't showing up here for some reason.

tych0 commented:
Sorry, I misunderstood what was going on here. Can you drop the merge commits, so the history is clean? This LGTM otherwise.

MagusLucius commented:
> Sorry, I misunderstood what was going on here. Can you drop the merge commits, so the history is clean? This LGTM otherwise.

![active_branches](https://user-images.githubusercontent.com/6981631/76371696-ad7cab80-6300-11ea-8e9b-d297716c26b2.png)

If I understand how to do that.  Can I simply click the trash bin next to #1 and #3?

tych0 commented:
Possibly, I don't know how to do it through the github interface. I'd do a
git rebase -i and squash all the commits together.

On Tue, Mar 10, 2020 at 6:56 PM MagusLucius <notifications@github.com>
wrote:

> Sorry, I misunderstood what was going on here. Can you drop the merge
> commits, so the history is clean? This LGTM otherwise.
>
> [image: active_branches]
> <https://user-images.githubusercontent.com/6981631/76371696-ad7cab80-6300-11ea-8e9b-d297716c26b2.png>
>
> If I understand how to do that. Can I simply click the trash bin next to
> #1 <https://github.com/qtile/qtile/issues/1> and #3
> <https://github.com/qtile/qtile/issues/3>?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/qtile/qtile/pull/1306?email_source=notifications&email_token=AAF7VV23WXBPB6FG6LMA7WTRG3OURA5CNFSM4HFB7IAKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEONXKXA#issuecomment-597390684>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAF7VV4EKHOL2TLL62OPTZTRG3OURANCNFSM4HFB7IAA>
> .
>

tych0 commented:
Ok, I went ahead and squashed this myself. I also added some docs about its deps and deleted the old mpd widget. Thanks for everyone's work on this!

tych0 commented:
Seems like this breaks our doc generations. I've tracked it to these three:

```       ('prepare_status', status_dict,
            'characters to show the status of MPD'),
        ('status_format', status_format, 'format for displayed song info.'),
        ('idle_format', idle_format,
            'format for status when mpd has no playlist.'),
```

Any chance you can take a look at fixing the doc generation?

m-col commented:
At some point the widget's mpd client has disconnected and it's not getting reconnected, so it's stopped updating. It should probably reconnect upon any action (click or poll) to continue updating.
